File: SmartReceiptsTests/Persistence/Helpers/DatabaseReceiptsDatesTest.m
//
//  DatabaseReceiptsDatesTest.m
//  SmartReceipts
//
//  Created by Jaanus Siim on 20/05/15.
//  Copyright (c) 2015 Will Baumann. All rights reserved.
//

#import <UIKit/UIKit.h>
#import <XCTest/XCTest.h>
#import "SmartReceiptsTestsBase.h"
#import "WBTrip.h"
#import "DatabaseTestsHelper.h"
#import "DatabaseTableNames.h"
#import "NSDate+Calculations.h"

@interface Database (Expose)

- (NSTimeInterval)maxSecondForReceiptsInTrip:(WBTrip *)trip onDate:(NSDate *)date;

@end

@interface DatabaseReceiptsDatesTest : SmartReceiptsTestsBase

@property (nonatomic, strong) WBTrip *testTrip;

@end

@implementation DatabaseReceiptsDatesTest

- (void)setUp {
    [super setUp];

    self.testTrip = [self.db createTestTrip];

    [self.db insertTestReceipt:@{ReceiptsTable.COLUMN_PARENT: self.testTrip, ReceiptsTable.COLUMN_DATE: [[NSDate date] dateByAddingDays:-1]}];
    [self.db insertTestReceipt:@{ReceiptsTable.COLUMN_PARENT: self.testTrip, ReceiptsTable.COLUMN_DATE: [[NSDate date] dateByAddingDays:-1]}];

    [self.db insertTestReceipt:@{ReceiptsTable.COLUMN_PARENT: self.testTrip, ReceiptsTable.COLUMN_DATE: [NSDate date]}];
    [self.db insertTestReceipt:@{ReceiptsTable.COLUMN_PARENT: self.testTrip, ReceiptsTable.COLUMN_DATE: [NSDate date]}];
    [self.db insertTestReceipt:@{ReceiptsTable.COLUMN_PARENT: self.testTrip, ReceiptsTable.COLUMN_DATE: [NSDate date]}];
}

- (void)testMaxSecondsYesterday {
    NSTimeInterval maxSeconds = [self.db maxSecondForReceiptsInTrip:self.testTrip onDate:[[NSDate date] dateByAddingDays:-1]];
    XCTAssertEqualWithAccuracy(2, maxSeconds, 0.1);
}

- (void)testMaxSecondsToday {
    NSTimeInterval maxSeconds = [self.db maxSecondForReceiptsInTrip:self.testTrip onDate:[NSDate date]];
    XCTAssertEqualWithAccuracy(3, maxSeconds, 0.1);
}

@end

File: SmartReceiptsTests/Persistence/Helpers/DatabaseReceiptsDatesTest.swift
//
//  DatabaseReceiptsDatesTest.swift
//  SmartReceiptsTests
//
//  Created by Bogdan Evsenev on 03/10/2018.
//  Copyright © 2018 Will Baumann. All rights reserved.
//

import XCTest
@testable import SmartReceipts

class DatabaseReceiptsDatesTest: SmartReceiptsTestsBase {
    
    // Defines the accuracy buffer for these tests to avoid failures related to slower build times
    let accuracyBuffer = TimeInterval(10)
    
    var testTrip: WBTrip!

    override func setUp() {
        super.setUp()
        testTrip = db.createTestTrip()
        db.insertTestReceipt([ReceiptsTable.Column.ParentId : self.testTrip, ReceiptsTable.Column.Date : NSDate().addingDays(-1)])
        db.insertTestReceipt([ReceiptsTable.Column.ParentId : self.testTrip, ReceiptsTable.Column.Date : NSDate().addingDays(-1)])
        
        db.insertTestReceipt([ReceiptsTable.Column.ParentId : self.testTrip, ReceiptsTable.Column.Date : NSDate()])
        db.insertTestReceipt([ReceiptsTable.Column.ParentId : self.testTrip, ReceiptsTable.Column.Date : NSDate()])
        db.insertTestReceipt([ReceiptsTable.Column.ParentId : self.testTrip, ReceiptsTable.Column.Date : NSDate()])
    }

    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
    }

    func testMaxSecondsYesterday() {
        let date = NSDate()
        let maxSeconds = db.maxSecondForReceipts(in: testTrip, on: date.addingDays(-1))
        XCTAssertEqual(date.secondsOfDay(), maxSeconds, accuracy: accuracyBuffer)
    }
    
    func testMaxSecondsToday() {
        let date = NSDate()
        let maxSeconds = db.maxSecondForReceipts(in: testTrip, on: date as Date)
        XCTAssertEqual(date.secondsOfDay(), maxSeconds, accuracy: accuracyBuffer)
    }

}
